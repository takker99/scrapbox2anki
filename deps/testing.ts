export * from "https://deno.land/std@0.186.0/testing/asserts.ts";
export * from "https://deno.land/std@0.186.0/testing/snapshot.ts";
