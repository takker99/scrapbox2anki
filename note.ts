/// <reference lib="deno.unstable" />

import {
  Block,
  CodeBlock,
  convertToBlock,
  Line,
  Node,
  packRows,
  parse,
  parseToRows,
  Table,
} from "./deps/scrapbox-parser.ts";
import { encodeTitleURI, toTitleLc } from "./deps/scrapbox.ts";
import { Line as BaseLine } from "./type.ts";
import { detectNoteTitle } from "./detectNoteTitle.ts";
import { parsePath, Path } from "./path.ts";
import { convert } from "./sb2html.ts";

/** 抽出したnote
 *
 * deckとnote typeの情報を参照できないため、 deno-anki の`Note`とは型が異なる
 */
export interface Note {
  guid: string;

  /** note ID */
  id: number;

  /** このnoteが属するdeckのデータが書き込まれているページへのパス */
  deck?: Path;

  /** このnoteが使うnote typeのデータが書き込まれているページへのパス */
  noteType?: Path;

  /** updated time of the note */
  updated: number;
  tags?: string[];

  /** field nameをキーとしたfield values
   *
   * unnamed fieldは`""`をキーとして格納する
   */
  fields: Map<string, string>;
}

export const parseNotes = (
  project: string,
  title: string,
  lines: BaseLine[],
): Note[] => {
  if (lines.length === 0) return [];
  const packs = packRows(
    parseToRows(lines.map((line) => line.text).join("\n")),
    { hasTitle: true },
  );

  const notes = new Map<
    string,
    Omit<Note, "deck" | "noteType" | "fields"> & {
      // scrapbox記法としてparseするfieldは`true`を入れる
      fields: Map<string, [boolean, string]>;
    }
  >();
  let deckRef: Path | undefined;
  let noteTypeRef: Path | undefined;
  /** 現在読んでいる`pack.rows[0]`の行番号 */
  let counter = 0;
  for (const pack of packs) {
    switch (pack.type) {
      case "title":
        counter++;
        break;
      case "table":
      case "line": {
        counter += pack.rows.length;
        if (deckRef && noteTypeRef) break;

        // iconがある行を調べる
        const block = convertToBlock(pack);
        const icons = getIcons(block);
        for (const icon of icons) {
          if (icon.toLowerCase().startsWith("deck-")) {
            deckRef ??= parsePath(icon, project);
          }
          if (icon.toLowerCase().startsWith("notetype-")) {
            noteTypeRef ??= parsePath(icon, project);
          }
        }
        break;
      }
      case "codeBlock": {
        counter += pack.rows.length;
        const codeBlock = convertToBlock(pack);
        if (codeBlock.type !== "codeBlock") throw SyntaxError();

        const noteTitle = detectNoteTitle(codeBlock.fileName);
        if (!noteTitle) break;
        const { guid, name, isScrapboxSyntax } = noteTitle;

        const note = notes.get(guid) ??
          {
            guid,
            id: Infinity,
            updated: -Infinity,
            fields: new Map<string, [boolean, string]>([[
              "SourceURL",
              [
                false,
                `https://scrapbox.io/${project}/${encodeTitleURI(title)}#${
                  lines[counter].id
                }`,
              ],
            ]]),
          };

        // note id は、コードブロックを構成する行の作成日時のうち、一番古いものを採用する
        note.id = Math.min(
          ...lines.slice(counter, counter + pack.rows.length).map((line) =>
            line.created * 1000
          ),
          note.id,
        );
        note.updated = Math.max(
          ...lines.slice(counter, counter + pack.rows.length).map((line) =>
            line.updated * 1000
          ),
          note.updated,
        );
        const [isSyntax, content] = note.fields.get(name) ?? [true, ""];
        note.fields.set(
          name,
          [
            isSyntax && isScrapboxSyntax,

            content ? `${content}\n${codeBlock.content}` : codeBlock.content,
          ],
        );

        notes.set(guid, note);
        break;
      }
    }
  }

  return [...notes.values()].map(({ fields: fieldsUnparsed, ...note }) => {
    // 構文解析しつつ、tagsを取り出す
    const tags: string[] = [];
    const dupCheck = new Set<string>();
    const crawlTag = (tag: string) => {
      const tagLc = toTitleLc(tag);
      if (dupCheck.has(tagLc)) return;
      dupCheck.add(tagLc);
      tags.push(tag);
    };
    const fields = new Map<string, string>();

    // 先にnote取得元project nameとpage titleをtagに入れておく
    crawlTag(project);
    crawlTag(title);

    for (
      const [guid, [isScrapboxSyntax, content]] of fieldsUnparsed.entries()
    ) {
      if (!isScrapboxSyntax) {
        fields.set(guid, content);
        continue;
      }
      const html = convert(
        parse(content, { hasTitle: false }) as (Table | CodeBlock | Line)[],
        project,
        crawlTag,
      );
      fields.set(guid, html);
    }

    return ({ ...note, deck: deckRef, noteType: noteTypeRef, fields, tags });
  });
};

/** Blockに含まれるiconを出現順にすべて取り出す */
const getIcons = (block: Block): string[] => {
  switch (block.type) {
    case "line":
      return block.nodes.flatMap((node) => getIconsFromNode(node));
    case "table":
      return block.cells.flatMap((row) =>
        row.flatMap((cell) => cell.flatMap((node) => getIconsFromNode(node)))
      );
    default:
      return [];
  }
};

const getIconsFromNode = (node: Node): string[] => {
  switch (node.type) {
    case "icon":
    case "strongIcon":
      return [node.path];
    case "decoration":
    case "quote":
      return node.nodes.flatMap((node) => getIconsFromNode(node));
    default:
      return [];
  }
};
