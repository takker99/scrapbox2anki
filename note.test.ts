import { assertSnapshot } from "./deps/testing.ts";
import json from "./sample-page.json" assert { type: "json" };
import { parseNotes } from "./note.ts";

Deno.test("parseNotes", async (t) => {
  const { title, lines } = json;
  await assertSnapshot(t, parseNotes("takker-dist", title, lines));
});
