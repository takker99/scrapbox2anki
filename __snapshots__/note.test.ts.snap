export const snapshot = {};

snapshot[`parseNotes 1`] = `
{
  deckRef: {
    project: "takker-dist",
    title: "deck-テスト用deck",
  },
  noteTypeRef: {
    project: "takker-dist",
    title: "NoteType-穴埋めサンプル",
  },
  notes: [
    {
      fields: [
        "1676880489000,63f32a681280f00000b3624d,1676880495000,対象によって{{c1::異なった価値判断の基準}}を使い分けることを{{c2::ダブルスタンダード::カタカナ英語}}と呼ぶ<br><br>#ダブルスタンダード",
        "https://scrapbox.io/takker-dist/%E3%83%80%E3%83%96%E3%83%AB%E3%82%B9%E3%82%BF%E3%83%B3%E3%83%80%E3%83%BC%E3%83%89#63f32d1e1280f00000962d88",
      ],
      guid: "%Y4u,z/)",
      id: 1676881183000,
      tags: [
        "ダブルスタンダード",
      ],
      updated: 1676881357000,
    },
    {
      fields: [
        "英語：{{c1::double standard}}<br>日本語：{{c2::二重規範}}",
        "https://scrapbox.io/takker-dist/%E3%83%80%E3%83%96%E3%83%AB%E3%82%B9%E3%82%BF%E3%83%B3%E3%83%80%E3%83%BC%E3%83%89#63f3295f1280f00000b36220",
      ],
      guid: ";)f\$&p&/",
      id: 1676880224000,
      tags: [
      ],
      updated: 1676880498000,
    },
    {
      fields: [
        "{{c1::二重規範::全部漢字}}は略して{{c2::ダブスタ}}とも呼ぶ",
        "https://scrapbox.io/takker-dist/%E3%83%80%E3%83%96%E3%83%AB%E3%82%B9%E3%82%BF%E3%83%B3%E3%83%80%E3%83%BC%E3%83%89#63f32aa91280f00000b362ad",
      ],
      guid: "b,shjX~Uk",
      id: 1676880554000,
      tags: [
      ],
      updated: 1676880688000,
    },
    {
      fields: [
        "2つ以上の異なった価値判断の基準を使う場合は、{{c1::二重規範}}ではなく{{c1::多重規範}}とも呼ぶ",
        "https://scrapbox.io/takker-dist/%E3%83%80%E3%83%96%E3%83%AB%E3%82%B9%E3%82%BF%E3%83%B3%E3%83%80%E3%83%BC%E3%83%89#63f32aaf1280f00000b362cd",
      ],
      guid: "[ZRe=axi",
      id: 1676880560000,
      tags: [
      ],
      updated: 1676880560000,
    },
    {
      fields: [
        "二重規範は主に人を{{c1::非難}}する場合に使われる",
        "https://scrapbox.io/takker-dist/%E3%83%80%E3%83%96%E3%83%AB%E3%82%B9%E3%82%BF%E3%83%B3%E3%83%80%E3%83%BC%E3%83%89#63f32ab51280f00000b362ed",
      ],
      guid: "s=y#TrE)",
      id: 1676880566000,
      tags: [
      ],
      updated: 1676880566000,
    },
    {
      fields: [
        "{{c1::ミオミオ}}は{{c2::父親}}に対して{{c2::ダブスタクソ親父}}と吐き捨てた",
        "https://scrapbox.io/takker-dist/%E3%83%80%E3%83%96%E3%83%AB%E3%82%B9%E3%82%BF%E3%83%B3%E3%83%80%E3%83%BC%E3%83%89#63f324571280f00000b361fa",
      ],
      guid: "F1odiPo{",
      id: 1676878936000,
      tags: [
      ],
      updated: 1676880665000,
    },
  ],
}
`;
